src/mesh_refiner.cpp

```diff
@@ -0,0 +1,243 @@
+#include <iostream>
+#include <fstream>
+#include <string.h>
+#include <dolfin.h>
+#include <ufc.h>
+#include "domain.h"
+#include "dirichlet.h"
+#include "mesh_refiner.h"
+extern "C" {
+  #include "fasp.h"
+  #include "fasp_functs.h"
+}
+
+#include "L2Error.h"
+#include "SemiH1error.h"
+#include "gradient_recovery.h"
+#include "poisson_cell_marker.h"
+
+//--------------------------------
+Mesh_Refiner::Mesh_Refiner (
+  const std::shared_ptr<const dolfin::Mesh> initial_mesh,
+  const std::size_t max_elements_in,
+  const std::size_t max_refine_depth_in,
+  const double entropy_per_cell
+) {
+  _mesh.reset(new const dolfin::Mesh(*initial_mesh));
+
+  _l2_form.reset(new L2Error::Functional(_mesh));
+  _semi_h1_form.reset(new SemiH1error::Functional(_mesh));
+
+  iteration = 0;
+  Mesh_Refiner::needs_to_solve = true;
+  Mesh_Refiner::needs_refinement = false;
+
+  Mesh_Refiner::max_elements = max_elements_in;
+  Mesh_Refiner::max_refine_depth = max_refine_depth_in;
+  Mesh_Refiner::entropy_tolerance_per_cell = entropy_per_cell;
+};
+//--------------------------------
+Mesh_Refiner::~Mesh_Refiner () {};
+//--------------------------------
+
+//--------------------------------
+std::shared_ptr<const dolfin::Mesh> Mesh_Refiner::get_mesh () {
+  return _mesh;
+};
+//--------------------------------
+std::shared_ptr<const dolfin::Mesh> Mesh_Refiner::multilevel_refinement (
+  std::vector<std::shared_ptr<const dolfin::Function>> entropy_potential_vector,
```

Comment:
it's a bad name for the term `\eta_i + q_i \phi`, which is used to compute the entropy:
`D_i exp(\eta_i) \nabla(\eta_i + q_i \phi)`.